docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+and eventually participate in consensus. The goal of StateSync is to
+facilitate setting up a new node as quickly as possible.
+
+## Considerations
+Because Tendermint doesn't know anything about the application state,
+the goal of StateSync of to broker messages between nodes and through
+the ABCI. The implementation will have multiple touch points through the
+codebase:
+
+* A StateSync reactor to facilitate peer communication
+* A Set of ABCI messages to transmit application state to the reactor
+* A Set of MultiStore APIs for exposing snapshot data to the ABCI
+* A Storage format with validation and performance conciderations
+
+### Implementation Properties
+Beyond the approach, any implementation of StateSync can be evaluated
+across different criteria:
+
+* Speed: Expected throughput of producing and consuming snapshots
+* Safety: Cost of pushing partial or completely invalid snapshots to a node
+* Effort: How much effort does an implementation require
+
+### Implementation Question
+
+* What is the format of a snapshot
+    * Complete snapshot 
+    * Ordered IAVL key ranges
+    * Compressed individually chunks which can be validated
+* How is data validated
+    * Trust a peer with it's data blindly
+    * Trust a majority of peers
+    * Use light client validation to validate any key based on consensus
+      produced merkle tree root
+* What are the performance characteristics
+    * Random vs sequential reads
+    * How parallelizeable is the scheduling algorithm
+
+### Proposals
+Broadly speaking there are two approaches to this problem which have had
+varying degrees of discussion and progress. These approach can be
+summarized as:
+
+* Lazy: Where snapshots are produced dynamically at request time
+* Eager: Where snapshots are produced periodically and served from disk at request time
+
+#### Lazy StateSync
+An [initial specification](https://docs.google.com/document/d/15MFsQtNA0MGBv7F096FFWRDzQ1vR6_dics5Y49vF8JU/edit?ts=5a0f3629) was published by Alexis Sellier.
+In this design, the state has a given `size` of primitive elements (like
+keys or nodes), each element is assigned a number from 0 to `size-1`,
+and chunks consists of a range of such elements.  Ackratos raised
+[some concerns](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit)
+about this design, somewhat specific to the IAVL tree, and mainly concerning
+performance of random reads and of iterating through the tree to determine element numbers
+(ie. elements aren't indexed by the element number).
+
+This design could be generalized to have numbered chunks instead of
+elements, where chunks are indexed by their chunk number and determined
+by some algorithm for grouping the elements in the tree. This would
+allow pre-computation and persistence of the chunks that could alleviate
+the concerns about real-time iteration and random reads.  However, there
+seems to be a larger issue with the "numbered chunks" approach: it does
```

Comment:
Maybe this sentence should be removed as in the meantime it turns out that this is possible.